airflow/providers/amazon/aws/triggers/s3.py

```diff
@@ -0,0 +1,100 @@
+# specific language governing permissions and limitations
+# under the License.
+from __future__ import annotations
+
+import asyncio
+from functools import cached_property
+from typing import Any, AsyncIterator, Callable
+
+from airflow.providers.amazon.aws.hooks.s3 import S3Hook
+from airflow.triggers.base import BaseTrigger, TriggerEvent
+
+
+class S3KeyTrigger(BaseTrigger):
+    """
+    S3KeyTrigger is fired as deferred class with params to run the task in trigger worker
+
+    :param bucket_name: Name of the S3 bucket. Only needed when ``bucket_key``
+        is not provided as a full s3:// url.
+    :param bucket_key:  The key being waited on. Supports full s3:// style url
+        or relative path from root level. When it's specified as a full s3://
+        url, please leave bucket_name as `None`.
+    :param wildcard_match: whether the bucket_key should be interpreted as a
+        Unix wildcard pattern
+    :param aws_conn_id: reference to the s3 connection
+    :param hook_params: params for hook its optional
+    :param check_fn: Function that receives the list of the S3 objects,
+        and returns a boolean
+    """
+
+    def __init__(
+        self,
+        bucket_name: str,
+        bucket_key: str | list[str],
+        wildcard_match: bool = False,
+        check_fn: Callable[..., bool] | None = None,
+        aws_conn_id: str = "aws_default",
+        poke_interval: float = 5.0,
+        **hook_params: Any,
+    ):
+        super().__init__()
+        self.bucket_name = bucket_name
+        self.bucket_key = bucket_key
+        self.wildcard_match = wildcard_match
+        self.check_fn = check_fn
+        self.aws_conn_id = aws_conn_id
+        self.hook_params = hook_params
+        self.poke_interval = poke_interval
+
+    def serialize(self) -> tuple[str, dict[str, Any]]:
+        """Serialize S3KeyTrigger arguments and classpath."""
+        return (
+            "airflow.providers.amazon.aws.triggers.s3.S3KeyTrigger",
+            {
+                "bucket_name": self.bucket_name,
+                "bucket_key": self.bucket_key,
+                "wildcard_match": self.wildcard_match,
+                "check_fn": self.check_fn,
+                "aws_conn_id": self.aws_conn_id,
+                "hook_params": self.hook_params,
+                "poke_interval": self.poke_interval,
+            },
+        )
+
+    @cached_property
+    def hook(self) -> S3Hook:
+        return S3Hook(aws_conn_id=self.aws_conn_id, verify=self.hook_params.get("verify"))
+
+    async def run(self) -> AsyncIterator[TriggerEvent]:
+        """Make an asynchronous connection using S3HookAsync."""
+        try:
+            async with self.hook.async_conn as client:
+                while True:
+                    if await self.hook.check_key_async(
+                        client, self.bucket_name, self.bucket_key, self.wildcard_match
+                    ):
+                        if self.check_fn is None:
```

Comment:
if this is the only usage you're going to make of it, passing it as a callable is misleading.

You should pass a boolean `get_files` instead or something like that